File: src/sine_wave_py/sine_wave_py/sine_wave_receiver.py
# -*- coding: utf-8 -*-
# Copyright 2025 ZiouHu
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import rclpy
from rclpy.node import Node
from rclpy.callback_groups import MutuallyExclusiveCallbackGroup
import cv2
import numpy as np
from cv_bridge import CvBridge
from sine_wave_cpp.msg import Signal
from sine_wave_cpp.srv import ConvertImage
from sine_wave_py.sine_wave_parameters import sine_wave_parameters


class SineWaveReceiver:
    """Class to receive sine wave data and provide image conversion service with dynamic parameter updates."""

    def __init__(self, node: Node):
        """
        Initialize the sine wave receiver.

        Args:
            node: ROS2 node
        """
        self.node = node
        self.param_listener = sine_wave_parameters.ParamListener(self.node)
        self.params = self.param_listener.get_params()  # get initial parameters

        # Get parameters
        self.amplitude = self.params.amplitude
        self.angular_frequency = self.params.angular_frequency
        self.phase = self.params.phase
        self.frequency = self.params.publisher_frequency

        # Validate parameters
        self.validate_parameters()

        # Create callback groups
        self.subscription_cb_group = MutuallyExclusiveCallbackGroup()
        self.service_cb_group = MutuallyExclusiveCallbackGroup()

        # Create subscription
        self.subscription = self.node.create_subscription(
            Signal,
            "sine_wave",
            self.sine_wave_callback,
            10,
            callback_group=self.subscription_cb_group,
        )

        # Create service
        self.service = self.node.create_service(
            ConvertImage,
            "convert_image",
            self.convert_image_service,
            callback_group=self.service_cb_group,
        )

        # Initialize bridge for OpenCV conversion
        self.bridge = CvBridge()

        # Create a timer for checking parameter updates (every 0.3 seconds)
        self.update_timer = self.node.create_timer(0.3, self.update_params_callback)

        self.node.get_logger().info(
            "SineWaveReceiver node with custom Service is ready."
        )

    def declare_parameters(self):
        """Declare all parameters."""
        self.node.declare_parameter("amplitude", 1.0)
        self.node.declare_parameter("angular_frequency", 1.0)
        self.node.declare_parameter("phase", 0.0)
        self.node.declare_parameter("publisher_frequency", 10.0)

    def validate_parameters(self):
        """Validate the parameters and set defaults if needed."""
        # Validate frequency
        if self.frequency < 0.001 or self.frequency > 99999.9:
            self.node.get_logger().error(
                f"Invalid publisher_frequency ({self.frequency}). "
                "Must be within [0.001, 99999.9]. Using fallback value 10.0 Hz."
            )
            self.frequency = 10.0

        # Validate amplitude
        if self.amplitude < 0.0000001 or self.amplitude > 99999.9:
            self.node.get_logger().error(
                f"Invalid amplitude ({self.amplitude}). "
                "Must be within [0.0000001, 99999.9]. Using fallback value 1.0."
            )
            self.amplitude = 1.0

        # Validate angular_frequency
        if self.angular_frequency < 0.000001 or self.angular_frequency > 99999.9:
            self.node.get_logger().error(
                f"Invalid angular_frequency ({self.angular_frequency}). "
                "Must be within [0.000001, 99999.9]. Using fallback value 1.0."
            )
            self.angular_frequency = 1.0

        # Validate phase
        if self.phase < -6.283184 or self.phase > 6.283184:
            self.node.get_logger().error(
                f"Invalid phase ({self.phase}). "
                "Must be within [-6.283184, 6.283184]. Using fallback value 0.0."
            )
            self.phase = 0.0

    def sine_wave_callback(self, msg):
        """
        Process the received sine wave message.

        Args:
            msg: Received Signal message
        """
        sine_value = msg.data
        stamp = msg.header.stamp
        self.node.get_logger().info(
            f"Received sine wave value: {sine_value}, timestamp: {stamp.sec}.{stamp.nanosec}"
        )

    def convert_image_service(self, request, response):
        """
        Service to convert a color image to grayscale.

        Args:
            request: Service request containing file path
            response: Service response to contain grayscale image

        Returns:
            Response with grayscale image
        """
        self.node.get_logger().info(
            f"Service callback invoked. file_path: {request.file_path}"
        )

        # Read the image using OpenCV
        color_img = cv2.imread(request.file_path, cv2.IMREAD_COLOR)
        if color_img is None:
            self.node.get_logger().error(
                f"Failed to read image from path: {request.file_path}"
            )
            return response

        self.node.get_logger().info("Image successfully read.")

        # Convert to grayscale
        gray_img = cv2.cvtColor(color_img, cv2.COLOR_BGR2GRAY)

        # Convert the image to a ROS Image message (encoding "mono8")
        response.grayscale_image = self.bridge.cv2_to_imgmsg(gray_img, encoding="mono8")

        self.node.get_logger().info("Successfully transformed image to gray image!")

        # Visualize the images side by side
        gray_bgr = cv2.cvtColor(gray_img, cv2.COLOR_GRAY2BGR)
        concatenated = np.hstack((color_img, gray_bgr))

        cv2.namedWindow("origin image | gray image", cv2.WINDOW_AUTOSIZE)
        cv2.imshow("origin image | gray image", concatenated)
        cv2.waitKey(0)
        cv2.destroyAllWindows()

        return response

    def update_params_callback(self):
        """
        Timer callback that checks for parameter updates.

        If new parameters are detected via the parameter listener, update the internal state.
        """
        if self.param_listener.is_old(self.params):
            # Retrieve new parameters
            new_params = self.param_listener.get_params()

            # Update internal state for publisher_frequency, amplitude, angular frequency, and phase
            self.frequency = new_params.publisher_frequency
            self.amplitude = new_params.amplitude
            self.angular_frequency = new_params.angular_frequency
            self.phase = new_params.phase

            # Store the new parameters as current parameters
            self.params = new_params

            self.node.get_logger().info(
                f"Parameters updated dynamically:\n"
                f"Publisher frequency: {self.frequency:.2f} Hz\n"
                f"Amplitude: {self.amplitude:.2f}\n"
                f"Angular frequency: {self.angular_frequency:.2f}\n"
                f"Phase: {self.phase:.2f}"
            )

File: src/sine_wave_cpp/src/sine_wave_receiver.cpp
/*
 * Copyright 2015 Ziou
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "sine_wave_cpp/sine_wave_receiver.hpp"

#include "sine_wave_cpp/sine_wave_parameters.hpp"

#include <cv_bridge/cv_bridge.h>  // NOLINT(build/include_order)

#include <mutex>
#include <thread>

#include <opencv2/opencv.hpp>
#include <rclcpp/rclcpp.hpp>
#include <rclcpp/service.hpp>

SineWaveReceiver::SineWaveReceiver(rclcpp::Node::SharedPtr node, const sine_wave::Params & params)
: node_(node),
  amplitude_(params.amplitude),
  angular_frequency_(params.angular_frequency),
  phase_(params.phase),
  frequency_(params.publisher_frequency)
{
  param_listener_ = std::make_shared<sine_wave::ParamListener>(node_);
  // validate frequency
  if (frequency_ < 0.001 || frequency_ > 99999.9) {
    RCLCPP_ERROR(
      node_->get_logger(),
      "Invalid publisher_frequency (%f). Must be within [0.001, 99999.9]. Using fallback value "
      "10.0 Hz.",
      frequency_);
    frequency_ = 10.0;
  }

  // Validate amplitude
  if (amplitude_ < 0.0000001 || amplitude_ > 99999.9) {
    RCLCPP_ERROR(
      node_->get_logger(),
      "Invalid amplitude (%f). Must be within [0.0000001, 99999.9]. Using fallback value 1.0.",
      amplitude_);
    amplitude_ = 1.0;
  }

  // Validate angular_frequency
  if (angular_frequency_ < 0.000001 || angular_frequency_ > 99999.9) {
    RCLCPP_ERROR(
      node_->get_logger(),
      "Invalid angular_frequency (%f). Must be within [0.000001, 99999.9]. Using fallback value "
      "1.0.",
      angular_frequency_);
    angular_frequency_ = 1.0;
  }

  // Validate phase
  if (phase_ < -6.283184 || phase_ > 6.283184) {
    RCLCPP_ERROR(
      node_->get_logger(),
      "Invalid phase (%f). Must be within [-6.283184, 6.283184]. Using fallback value 0.0.",
      phase_);
    phase_ = 0.0;
  }

  // define the callbackgroup to introduce the multi-thread, use MutuallyExclusive to make
  // subscription to be extra thread
  auto subscription_cb_group =
    node_->create_callback_group(rclcpp::CallbackGroupType::MutuallyExclusive);
  auto service_cb_group =
    node_->create_callback_group(rclcpp::CallbackGroupType::MutuallyExclusive);

  // create subscription
  rclcpp::SubscriptionOptions sub_options;
  sub_options.callback_group = subscription_cb_group;
  subscription_ = node_->create_subscription<sine_wave_cpp::msg::Signal>(
    "sine_wave", 10, std::bind(&SineWaveReceiver::sinewaveCallback, this, std::placeholders::_1),
    sub_options);

  // create Service
  service_ = node_->create_service<sine_wave_cpp::srv::ConvertImage>(
    "convert_image",
    std::bind(
      &SineWaveReceiver::convertImageService, this, std::placeholders::_1, std::placeholders::_2),
    rmw_qos_profile_default);

  RCLCPP_INFO(node_->get_logger(), "SineWaveReceiver node with custom Service is ready.");
}

void SineWaveReceiver::sinewaveCallback(const sine_wave_cpp::msg::Signal::SharedPtr msg)
{
  double sine_value = msg->data;
  auto stamp = msg->header.stamp;
  RCLCPP_INFO(
    node_->get_logger(), "Received sine wave value: %f, timestamp: %d.%u", sine_value, stamp.sec,
    stamp.nanosec);
}

void SineWaveReceiver::convertImageService(
  const std::shared_ptr<sine_wave_cpp::srv::ConvertImage::Request> request,
  std::shared_ptr<sine_wave_cpp::srv::ConvertImage::Response> response)
{
  // 1. use opencv to read image
  RCLCPP_INFO(
    node_->get_logger(), "Service callback invoked. file_path: %s", request->file_path.c_str());

  cv::Mat color_img = cv::imread(request->file_path, cv::IMREAD_COLOR);
  if (color_img.empty()) {
    RCLCPP_ERROR(
      node_->get_logger(), "Failed to read image from path: %s", request->file_path.c_str());
    response->grayscale_image = sensor_msgs::msg::Image();
    return;
  }

  RCLCPP_INFO(node_->get_logger(), "Image successfully read.");

  // 2. to gary image
  cv::Mat gray_img;
  cv::cvtColor(color_img, gray_img, cv::COLOR_BGR2GRAY);

  // 3. transform it to sensor_msgs::Image and encoding to "mono8"
  std_msgs::msg::Header header;
  header.stamp = node_->now();
  auto cv_bridge_img = cv_bridge::CvImage(header, "mono8", gray_img);
  sensor_msgs::msg::Image ros_img_msg = *cv_bridge_img.toImageMsg();

  // 4. set the response
  response->grayscale_image = ros_img_msg;

  RCLCPP_INFO(node_->get_logger(), "Successfully transform image to gray image!");

  // 5. visualize it
  cv::Mat gray_bgr;
  cv::cvtColor(gray_img, gray_bgr, cv::COLOR_GRAY2BGR);

  cv::Mat concatenated;
  cv::hconcat(color_img, gray_bgr, concatenated);

  cv::namedWindow("oringin image | gray image", cv::WINDOW_AUTOSIZE);
  cv::imshow("oringin image | gray image", concatenated);
  cv::waitKey(0);
  cv::destroyAllWindows();
  // {
  //   std::lock_guard<std::mutex> lock(g_img_mutex);
  //   concatenated.copyTo(g_img_to_display);
  //   g_new_image_available = true;
  // }
  // g_img_cv.notify_one();
}

void SineWaveReceiver::updateParamsCallback()
{
  // Check if parameters have been updated via the parameter listener
  if (param_listener_->is_old(params_)) {
    // Retrieve new parameters
    auto new_params = param_listener_->get_params();

    // Update internal state for publisher_frequency, amplitude, angular frequency, and phase
    frequency_ = new_params.publisher_frequency;
    amplitude_ = new_params.amplitude;
    angular_frequency_ = new_params.angular_frequency;
    phase_ = new_params.phase;

    // Store the new parameters as current parameters
    params_ = new_params;
    // RCLCPP_INFO(
    //   node_->get_logger(),
    //   "Parameters updated dynamically:\n"
    //   "Publisher frequency: %.2f Hz\n"
    //   "Amplitude: %.2f\n"
    //   "Angular frequency: %.2f\n"
    //   "Phase: %.2f",
    //   frequency_, amplitude_, angular_frequency_, phase_);
  }
}
